source/NVDAObjects/window/winword.py

```diff
@@ -1306,6 +1409,57 @@ def script_toggleAlignment(self,gesture):
 		if msg:
 			ui.message(msg)
 
+	@script(gestures=["kb:control+shift+a", "kb:control+shift+k"])
+	def script_toggleCaps(self, gesture):
+		if not self.WinwordSelectionObject:
+			# We cannot fetch the Word object model, so we therefore cannot report the format change.
+			# The object model may be unavailable because this is a pure UIA implementation such as Windows 10 Mail,
+			# or its within Windows Defender Application Guard.
+			# Eventually UIA will have its own way of detecting format changes at the cursor.
+			# For now, just let the gesture through and don't report anything.
+			return gesture.send()
+		val = self._WaitForValueChangeForAction(
+			lambda: gesture.send(),
+			lambda: (self.WinwordSelectionObject.font.allcaps, self.WinwordSelectionObject.font.smallcaps)
+		)
+		if val[0]:
+			# Translators: a message when toggling formatting to 'all capital' in Microsoft word
+			ui.message(_("All caps on"))
+		elif val[1]:
+			# Translators: a message when toggling formatting to 'small capital' in Microsoft word
+			ui.message(_("Small caps on"))
+		else:
+			# Translators: a message when toggling formatting to 'No capital' in Microsoft word
+			ui.message(_("Caps off"))
+
+	@script(gesture="kb:shift+f3")
+	def script_changeCase(self, gesture):
+		if not self.WinwordSelectionObject:
+			# We cannot fetch the Word object model, so we therefore cannot report the format change.
+			# The object model may be unavailable because this is a pure UIA implementation such as Windows 10 Mail,
+			# or its within Windows Defender Application Guard.
+			# For now, just let the gesture through and don't report anything.
+			return gesture.send()
+		val = self._WaitForValueChangeForAction(
+			lambda: gesture.send(),
+			lambda: self.WinwordSelectionObject.Range.Case
+		)
+		# Under Outlook, calling the script quickly a second time gives strange results.
+		# Case value passes sometimes through an intermediate value.
+		# So poll the value a second time.

```

Comment:
It may be that the first operation hasn't yet completed. It would probably be best if nvda scripts queued subsequent runs until the first is complete. Could you see if you can confirm if that is happening here?

Reply:
How do you perform this? I.e. how do you delay the second script until the end of the first?

Moreover, I have tried to reproduce the issue with only one polling, but I cannot anymore. I will try on another version of Word later.

Reply:
I have just tested on another Word and Outlook version and I cannot reproduce the issue either. Both Office are 2016, but one of them was not up-to-date.

Reply:
>How do you perform this? I.e. how do you delay the second script until the end of the first?

We'd need to first confirm that is what is happening, my suggestion was really just a guess. Since I didn't know, I have done some investigation, the short version is that it seems unlikely that the script is being called again before it completes. The long answer follows:

To work it out we can look at how the gestures are called. 

It all starts with the `hookThreadFunc` in `source/winInputHook.py`. This runs in it's own thread, it registers a Low level hook function for the keyboard. Unfortunately this is where the questions start: 
- Is this callback `keyboardHook` called only in the context of the thread that installed it? 
  - The [docs for "LowLevelKeyboardProc callback function"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644985(v=vs.85)) say:
  > This hook is called in the context of the thread that installed it. The call is made by sending a message to the thread that installed the hook. Therefore, the thread that installed the hook must have a message loop.
- But the ["Hooks Overview" docs ](https://docs.microsoft.com/en-us/windows/win32/winmsg/about-hooks) say the following, please note that [`WH_KEYBOARD_LL` is a "global only" install](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw):
 >  A global hook procedure can be called in the context of any application in the same desktop as the calling thread, so the procedure must be in a separate DLL module. 
- Can this callback be re-entered by the Windows system?
  - I couldn't find any information about this. 

Another interesting piece of advice from the [docs for "LowLevelKeyboardProc callback function"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644985(v=vs.85)) which we are apparently ignoring is:
>Note  Debug hooks cannot track this type of low level keyboard hooks. If the application must use low level hooks, it should run the hooks on a dedicated thread that passes the work off to a worker thread and then immediately returns. In most cases where the application needs to use low level hooks, it should monitor raw input instead. This is because raw input can asynchronously monitor mouse and keyboard messages that are targeted for other threads more effectively than low level hooks can. For more information on raw input, see Raw Input.

Ok, well I say "ignoring", but it looks like most paths result in adding an event to the event queue. The logic is complicated enough that it is hard to confirm this happens for all paths. If we assume it does happen for all paths, then I think we can be fairly confident that the script can't be called until it is exited.

I'm going to call on @michaelDCurran to take a look at this to see if there is anything I might have missed here, or if he can clarify anything further.

Reply:
@feerrenrut thanks for your detailed explanations (and also for the summary!).

I have tried again but I cannot reproduce the issue anymore, This PR is quite old (it was one of my first PRs) and I do not remember exactly which type of stress test may have produced the issue; I also do not remember for example if I had tried to restart my computer.
I have thus removed this double polling. If an issue occur we my investigate on this new basis.

Reply:
@feerrenrut, have you had the opportunity to discuss with @michaelDCurran on the double script call potential issue.
Since I have removed the double-polling anyway, this conversation is not blocked anymore IMO. Do you agree?